wallet: Disable creating and loading legacy wallets

achow101 opened the pull request:
To prepare for the deletion of legacy wallet code, disable creating or loading new legacy wallets.

Tests for the legacy wallet specifically are deleted.

Split from https://github.com/bitcoin/bitcoin/pull/28710

achow101 committed:
test: Remove legacy wallet tests from wallet_reindex.py

achow101 committed:
test: Remove legacy wallet unit tests

achow101 committed:
test: wallet_signer.py bdb will be removed

achow101 committed:
test: Remove legacy wallet tests from wallet_backwards_compatibility.py

achow101 committed:
test: remove legacy wallet functional tests

Removes all legacy wallet specific functional tests.

Also removes the --descriptor and --legacy-wallet options as these are
no longer necessary with the legacy wallet removed.

achow101 committed:
wallet: Remove -format and bdb from wallet tool's createfromdump

achow101 committed:
wallet: Remove wallettool salvage

Salvage is bdb only which is about to be removed.

achow101 committed:
bench: Remove WalletLoadingLegacy benchmark

achow101 committed:
wallet: Disallow loading legacy wallets

Legacy wallets do not have the descriptors flag set. Don't load wallets
without the descriptors flag.

At the same time, we will no longer load BDB databases since they are
only used for legacy wallets.

achow101 commented:
Marking this as ready for review as it no longer has any prerequisites.

maflcko reviewed (approved):
Needs rebase.

I left some nits, feel free to ignore.

I skipped over the large commit and the last two.

Otherwise:

review ACK 82fc65c9ab53c416935b5eb20be10fe69ded8b88 👨

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: review ACK 82fc65c9ab53c416935b5eb20be10fe69ded8b88 👨
1738Jue8VRvURHJG6ZrNb0UzewDcwNZT7SO5vi0tlLfJlYOf/7IVJSk5W9mompg0cCbyr9xc3mF6TE876MUeBg==
```

</details>

achow101 commented on src/qt/test/wallettests.cpp:
```diff
@@ -395,56 +376,6 @@ void TestGUI(interfaces::Node& node, const std::shared_ptr<CWallet>& wallet)
 
-void TestGUIWatchOnly(interfaces::Node& node, TestChain100Setup& test)
-{
-    const std::shared_ptr<CWallet>& wallet = SetupLegacyWatchOnlyWallet(node, test);
```
No, watch-only descriptor wallets do not change the GUI as watch-only legacy wallets did.

achow101 replied:
Added a commit to make the test work for descriptor wallets.

achow101 commented on test/functional/wallet_transactiontime_rescan.py:
```diff
@@ -144,20 +144,6 @@ def run_test(self):
         restorewo_wallet.importaddress(wo2, rescan=False)
         restorewo_wallet.importaddress(wo3, rescan=False)
 
-        # check user has 0 balance and no transactions
```
`importaddress` calls `importdescriptors` in the test framework. These checks are that `importaddress` (the legacy wallet RPC) doesn't rescan and will require a separate rescan. But that will not work once the tests is descriptor wallets only.

Will expand the commit message.

davidgumberg replied:
AFAICT the workaround is correct, and this test is already using `importdescriptors` on master.

This test succeeds as-is on this branch with this commit (https://github.com/bitcoin/bitcoin/pull/31250/commits/e6c900f9c0e1b143098663e1c5c80e1811946b2d) dropped.

achow101 replied:
I certain that this commit was required when writing this PR originally, however, I no longer recall the error that it fixed. Dropping this commit.

achow101 commented on CMakeLists.txt:
```diff
@@ -626,7 +624,6 @@ message("  libbitcoinkernel (experimental) ..... ${BUILD_KERNEL_LIB}")
 message("  wallet support ...................... ${ENABLE_WALLET}")
 if(ENABLE_WALLET)
-  message("   - descriptor wallets (SQLite) ...... ${WITH_SQLITE}")
   message("   - legacy wallets (Berkeley DB) ..... ${WITH_BDB}")
```
It only removes sqlite as an optional dependency when the wallet is enabled, so I think having this log line is redundant. Removing the BDB line is still incorrect at this time as it can be built with and used (in this commit).

Sjors replied:
Done in #31961

achow101 commented on src/bench/wallet_ismine.cpp:
```diff
@@ -70,11 +70,6 @@ static void WalletIsMine(benchmark::Bench& bench, bool legacy_wallet, int num_co
-#endif
-
 #ifdef USE_SQLITE
 static void WalletIsMineDescriptors(benchmark::Bench& bench) { WalletIsMine(bench, /*legacy_wallet=*/false); }
```
Done

achow101 commented on test/functional/wallet_backwards_compatibility.py:
Done

achow101 commented on test/functional/wallet_reindex.py:
```diff
@@ -15,12 +15,10 @@
     def set_test_params(self):
         self.num_nodes = 1
         self.setup_clean_chain = True
+        self.uses_wallet = True
```
Removed

achow101 commented on src/wallet/dump.cpp:
```diff
@@ -175,25 +175,12 @@ bool CreateFromDump(const ArgsManager& args, const std::string& name, const fs::
+    // Other formats such as BDB should not be loaded into a sqlite database since they also
+    // use a different type of wallet entirely which is no longer compatible with this software.
+    if (format_value != "sqlite") {
+        error = _("Error: Dumpfile specifies an unsupported database format");
```
Done

maflcko reviewed (approved):
Clarified my nits a bit, nothing blocking.

I still haven't reviewed the last commit and the large commit. I wonder if it is possible to split the large commit into one that removes test files (or test cases) and one that does the test framework change, possibly making review easier.

re-ACK b8494154e174481bbc0350013f1960973f35f3fc 🔑

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: re-ACK b8494154e174481bbc0350013f1960973f35f3fc 🔑
FHB3gaGmlXUnVYceoczXV52txwmCFbqMqxMuJLOB1MC2BCTd8xLZoCVRxUH3Kz+37u17PYG18Jd3OynL2tgbDg==
```

</details>

Sjors commented:
At first glance it makes sense to make 189eb8c02b5b81fe3239df34cd9ef48871e2e5de its own PR, i.e. making sqlite mandatory for wallet building. See https://github.com/bitcoin/bitcoin/pull/31250#discussion_r1963117438

brunoerg commented on test/functional/test_runner.py:
```diff
@@ -86,7 +86,6 @@
     'feature_pruning.py',
     'feature_dbcrash.py',
     'feature_index_prune.py',
-    'wallet_pruning.py --legacy-wallet',
```
56e9bef655673becb3e265948143a9e860a5c68b: nit: could also update the example in `test_runner.py`:
```py
for exclude_test in exclude_tests:
    # A space in the name indicates it has arguments such as "wallet_basic.py --descriptors"
    if ' ' in exclude_test:
```

achow101 replied:
Done

brunoerg commented on src/wallet/wallet.cpp:
```diff
@@ -3049,7 +3049,11 @@ std::shared_ptr<CWallet> CWallet::Create(WalletContext& context, const std::stri
             return nullptr;
-        } else {
+        } else if (nLoadWalletRet == DBErrors::LEGACY_WALLET) {
+            error = strprintf(_("Error loading %s: Wallet is a legacy wallet. Please migrate to a descriptor wallet using the migration tool (migratewallet RPC)."), walletFile);
```
ba2042c31f1630678a1f08bb68e01c1d98cc9abc: Is this verification reachable? The scenario I can see is: someone unloaded the wallet using an old version, upgraded and then tried to load it again. However, before reaching here it would fail during the database creation (`MakeDatabase`).

achow101 replied:
There is a theoretical legacy-sqlite wallet configuration that could reach this. This is actually an unsupported configuration, although it technically works today.

brunoerg replied:
Got it, thanks.

Sjors reviewed (approved):
ACK ba2042c31f1630678a1f08bb68e01c1d98cc9abc

Though there's a few places where we could salvage test coverage (in a followup).

I ran the test suite on all commits.

`wallet_upgradewallet.py` is dropped. The test itself was only run for legacy wallets. However the `upgradewallet` RPC isn't legacy-only, even though it doesn't (currently) do anything useful for descriptor wallets. Maybe we should drop it here or in #28710? Or otherwise at least have some trivial coverage.

Should `test/functional/wallet_pruning.py` be retrofitted for descriptor wallets instead of deleting it?

First commit now matches ac6cde731314d981391bc313c98d671c68211d33 from #31961.

Note that 928760e6dd8383db2dbe87f99185b07cf13bcab3 drops backward compatibility tests for the legacy wallet, but ba2042c31f1630678a1f08bb68e01c1d98cc9abc brings new coverage (and we also have dedicated migration coverage).

brunoerg commented on test/functional/wallet_backwards_compatibility.py:
```diff
@@ -363,27 +294,52 @@ def run_test(self):
+            pubkey = addr_info["pubkey"]
+
+            # Make a backup of the wallet file
+            backup_path = os.path.join(self.options.tmpdir, f"{wallet_name}.dat")
```
ba2042c31f1630678a1f08bb68e01c1d98cc9abc: both `backup_path` and `pubkey` are unused.

achow101 replied:
`backup_path` is used, but `pubkey` and `hdkeypath` are not. Removed those.

brunoerg reviewed (approved):
code review ACK ba2042c31f1630678a1f08bb68e01c1d98cc9abc 
no blockers!

- 8afd005cbf9209e90a5cc1bcdb2c0ff400c2de33: checked it removes disable RPCs related to legacy wallets (import/dump stuff) - I had same doubt of https://github.com/bitcoin/bitcoin/pull/31250/commits/8afd005cbf9209e90a5cc1bcdb2c0ff400c2de33#r1985126112
- 56e9bef655673becb3e265948143a9e860a5c68b: checked it removes legacy wallet functional tests and get rid off --descriptor --legacy-wallet. Left a nit about a comment that mentions --descriptor as example that could be removed.
- 1cafc80e87d98f43bf00f8663aa96d598085c1d8: checked that `restorewo_wallet` is a descriptor wallet (can be confirmed with `getwalletinfo` - `{'walletname': 'wo', 'walletversion': 169900, 'format': 'sqlite', 'balance': Decimal('0E-8'), 'unconfirmed_balance': Decimal('0E-8'), 'immature_balance': Decimal('0E-8'), 'txcount': 0, 'keypoolsize': 0, 'keypoolsize_hd_internal': 0, 'paytxfee': Decimal('0E-8'), 'private_keys_enabled': False, 'avoid_reuse': False, 'scanning': False, 'descriptors': True, 'external_signer': False, 'blank': False, 'lastprocessedblock': {'hash': '77a656f562fcce93cb25747e1d3ef773b0803feca410587bd3e6cff9975dd5fc', 'height': 611}}`), so there is no need to rescan since `importaddress` will call `importdescriptors` and therefore do it.
- 52c2774269d568c92d5c65b37c8ce32c745edfe7: since we cannot create or load legacy wallet, seems fine to remove salvage. Maybe we could also remove comments that mention it? (e.g. "Try using the wallet tool bitcoin-wallet to salvage or restoring a backup." - src/wallet/bdb.cpp)?

Successfully ran bench, unit and functional tests on MacOS 14.3.

Sjors commented:
Hoorway, #31961 landed.

re-ACK 32a522f2d825957f0c85d7b4ea9185a053b018e3

laanwj commented on src/bitcoin-wallet.cpp:
```diff
@@ -40,13 +40,11 @@ static void SetupWalletToolArgs(ArgsManager& argsman)
 
     argsman.AddCommand("info", "Get wallet info");
     argsman.AddCommand("create", "Create new wallet file");
-    argsman.AddCommand("salvage", "Attempt to recover private keys from a corrupt wallet. Warning: 'salvage' is experimental.");
```
At some point we might want to have something like `-salvage` for corrupted sqlite wallets, but makes sense to remove it for now.

brunoerg reviewed (approved):
reACK 32a522f2d825957f0c85d7b4ea9185a053b018e3

laanwj commented on src/wallet/dump.cpp:
```diff
@@ -175,25 +175,12 @@ bool CreateFromDump(const ArgsManager& args, const std::string& name, const fs::
+    // Other formats such as BDB should not be loaded into a sqlite database since they also
+    // use a different type of wallet entirely which is no longer compatible with this software.
+    if (format_value != "sqlite") {
+        error = strprintf(_("Error: Dumpfile specifies an unsupported database format (%s). Only sqlite database dumps are supported"), format_value);
```
missing `return false;`?

achow101 replied:
Oops yes, done.

laanwj commented on src/wallet/wallet.cpp:
```diff
@@ -3052,7 +3052,11 @@ std::shared_ptr<CWallet> CWallet::Create(WalletContext& context, const std::stri
+        } else if (nLoadWalletRet == DBErrors::LEGACY_WALLET) {
+            error = strprintf(_("Error loading %s: Wallet is a legacy wallet. Please migrate to a descriptor wallet using the migration tool (migratewallet RPC)."), walletFile);
+            return nullptr;
+        }
```
nit: don't we normally use same-line `} else {`?

achow101 replied:
Done

laanwj commented on src/wallet/walletdb.cpp:
```diff
@@ -485,6 +485,11 @@ static DBErrors LoadWalletFlags(CWallet* pwallet, DatabaseBatch& batch) EXCLUSIV
         }
+        // All wallets must be descriptor wallets unless opened with a bdb_ro db
+        // bdb_ro is only used for legacy to descriptor migration.
+        if (pwallet->GetDatabase().Format() != "bdb_ro" && !pwallet->IsWalletFlagSet(WALLET_FLAG_DESCRIPTORS)) {
```
Why do we compare strings here instead of use the `enum DatabaseFormat`?

pablomartin4btc reviewed (commented):
Concept ACK

Light cr (by commit - will do another round next week) & tACK 32a522f2d825957f0c85d7b4ea9185a053b018e3

Verified when I try to load a legacy wallet (via RPC or from `QT` menu) I get: 
"_`... The wallet appears to be a Legacy wallet, please use the wallet migration tool (migratewallet RPC).`_"

Also, running `QT` with legacy wallets in the `settings.json` will produce same output above closing the app with that error description ("_`Error: Failed to open database path...`_").

I can see we are still able to create legacy via RPC (n longer the case in final #28710 - with `-deprecatedrpc=create_bdb`), and if I run te command in the QT console I get the wallet loaded:
 
![image](https://github.com/user-attachments/assets/54291a7c-2631-4cd3-bec7-e9d0db6c8529)

Is this correct or needs to be fixed in this PR?

pablomartin4btc reviewed (commented):
Adding more details from my previous [comment](https://github.com/bitcoin/bitcoin/pull/31250#pullrequestreview-2705644380):
- The legacy wallet is being created as a `sqlite` DB (the unsupported legacy-sqlite mentioned [above](https://github.com/bitcoin/bitcoin/pull/31250#discussion_r1983725941)):
https://github.com/bitcoin/bitcoin/blob/32a522f2d825957f0c85d7b4ea9185a053b018e3/src/wallet/walletdb.cpp#L1471-L1477
- In `getwalletinfo` we could see `"format": "sqlite",`;

achow101 commented:
Rebased and addressed comments.

Will look into the qt issue tomorrow.

pablomartin4btc commented:
> Will look into the qt issue tomorrow.

I was using `QT` for testing it but I think the issue is happening at the node level.

achow101 committed:
wallet: Disallow creating legacy wallets

Remove the option to set descriptors=False when creating a wallet, and
enforce this in RPC and in CreateWallet

achow101 commented:
The issue with creating a legacy-sqlite wallet should be fixed now.

laanwj reviewed (approved):
Code review ACK 545a36ae0aa5687366fc4fdd54058072f333e73f 
Assuming this is still planned for 30.0, it's probably good to merge the BDB removal early in the release window, so that new PRs no longer need to take both cases into account, and possible problems can come to light before the release.

pablomartin4btc commented on test/functional/wallet_descriptor.py:
```diff
@@ -148,18 +148,6 @@ def run_test(self):
         send_wrpc.sendtoaddress(addr, 10)
 
-        # Make sure things are disabled
-        self.log.info("Test disabled RPCs")
```
I second this, e.g. `importprivkey` and other rpc are not removed until #28710, just curious what's the rationale to remove the calls from here and not from other places?

achow101 replied:
Moved this commit to #28710

pablomartin4btc reviewed (commented):
tACK 545a36ae0aa5687366fc4fdd54058072f333e73f

<details>
<summary>The <a href="https://github.com/bitcoin/bitcoin/pull/31250#pullrequestreview-2705644380">issue</a> I've described previously has been fixed.</summary>

```
./build_3125/bin/bitcoind -regtest -deprecatedrpc="create_bdb"
...
```
(on another terminal)
```
./build_31250/bin/bitcoin-cli -regtest createwallet "legacy_test" true true "" false false
error code: -4
error message:
descriptors argument must be set to "true"; It is no longer possible to create a legacy wallet.

```

</details>

Left a couple of comments.

achow101 committed:
tests, gui: Use descriptors watchonly wallet for watchonly test

achow101 commented on src/wallet/test/walletload_tests.cpp:
```diff
@@ -83,132 +83,5 @@ BOOST_FIXTURE_TEST_CASE(wallet_load_descriptors, TestingSetup)
-}
-
-
-BOOST_FIXTURE_TEST_CASE(wallet_load_ckey, TestingSetup)
```
This test case is no longer interesting because it is related to auto upgrade behavior for legacy wallets that will not occur for descriptor wallets.

achow101 commented on src/wallet/test/wallet_tests.cpp:
```diff
@@ -195,141 +195,6 @@ BOOST_FIXTURE_TEST_CASE(scan_for_wallet_transactions, TestChain100Setup)
     }
 }
 
-BOOST_FIXTURE_TEST_CASE(importmulti_rescan, TestChain100Setup)
```
Yes. I don't think having this unit test is all that useful.

achow101 commented on test/functional/wallet_backwards_compatibility.py:
```diff
@@ -249,28 +206,23 @@ def run_test(self):
-                node.rpc.createwallet(wallet_name=wallet_name, descriptors=self.options.descriptors)
-            else:
-                node.rpc.createwallet(wallet_name=wallet_name)
+            node.rpc.createwallet(wallet_name=wallet_name, descriptors=True)
```
Done

achow101 replied:
Actually, no. This parameter needs to be here because `node` includes older versions where descriptors is not the default.

pablomartin4btc replied:
But those are descriptors nodes already (`for node in descriptors_nodes:`).

achow101 replied:
Those are nodes where descriptor wallets are available, not when they were the default. We did not default to descriptor wallets until a couple of releases after introduction.

pablomartin4btc replied:
Ok I see, checked the comment in the code, thanks!

rkrux reviewed (commented):
Concept ACK 380f5e2359447ed2e9a1b54cffbefef1eb4f9b6a

I've looked at the last two commits only wherein creation of legacy wallets is disallowed and loading of legacy wallets is disallowed respectively. 

For the latter, the `BERKELEY` database format is removed & `MakeDatabase` instead defaults to using the `BERKELEY_RO` database format, before loading the wallet, while ensuring that `BERKELEY_RO` can only be used during the wallet migration flow wherein the `require_format` is explicitly set to `BERKELEY_RO` already in the migration flow - this ensures that `legacy-bdb` wallets are not loaded.

However, a theoretical `legacy-sqlite` wallet can still bypass this^, that is later prohibited to be loaded while checking for legacy non `bdb_ro` wallets during loading the wallet flags.

Thus, I believe both legacy bdb and sqlite wallets are disallowed to be loaded with the exclusion of loading/using legacy wallets only during the migration flow.

rkrux commented on src/wallet/rpc/wallet.cpp:
```diff
@@ -402,13 +400,9 @@ static RPCHelpMan createwallet()
+        throw JSONRPCError(RPC_WALLET_ERROR, "descriptors argument must be set to \"true\"; It is no longer possible to create a legacy wallet.");
     }
     if (!request.params[7].isNull() && request.params[7].get_bool()) {
 #ifdef ENABLE_EXTERNAL_SIGNER
```
I see it's removed later in #28710 but can be removed here as well as I feel it has become redundant in this PR itself.

achow101 replied:
Removed

achow101 commented on src/wallet/rpc/wallet.cpp:
```diff
@@ -355,9 +355,7 @@ static RPCHelpMan createwallet()
-            {"descriptors", RPCArg::Type::BOOL, RPCArg::Default{true}, "Create a native descriptor wallet. The wallet will use descriptors internally to handle address creation."
-                                                                       " Setting to \"false\" will create a legacy wallet; This is only possible with the -deprecatedrpc=create_bdb setting because, the legacy wallet type is being deprecated and"
-                                                                       " support for creating and opening legacy wallets will be removed in the future."},
+            {"descriptors", RPCArg::Type::BOOL, RPCArg::Default{true}, "If set, must be \"true\""},
```
The main reason is for backwards compatibility with older scripts, `bitcoin-cli`, etc. Typically, we don't ever remove positional arguments, and instead turn them into dummy arguments that must be set to a particular value. This PR doesn't rename it to dummy yet since I had some issues with doing that because of the backwards compatibility test.

achow101 commented on test/functional/wallet_backwards_compatibility.py:
```diff
@@ -285,5 +285,35 @@ def run_test(self):
+                self.restart_node(node.index, extra_args=[f"-wallet={wallet_name}"])
+                wallet_prev = node.get_wallet_rpc(wallet_name)
+                address = wallet_prev.getnewaddress('', "bech32")
+                addr_info = wallet_prev.validateaddress(address)
```
Removed

achow101 commented on src/wallet/walletdb.cpp:
```diff
@@ -1478,9 +1485,6 @@ std::unique_ptr<WalletDatabase> MakeDatabase(const fs::path& path, const Databas
     // If the format is not specified or detected, choose the default format based on what is available. We prefer BDB over SQLite for now.
```
Removed

Sjors commented:
re-ACK 1ffea3fba3b22c40bbcc44cd7c3a637e0450315a

Main changes since my last review:
- commit 3841da0f62fa5e26f679a12c6efbafe1cb17c25f to preserve PSBT GUI coverage
- #32214 landed which made 3841da0f62fa5e26f679a12c6efbafe1cb17c25f and 1ffea3fba3b22c40bbcc44cd7c3a637e0450315a smaller
- #31866 landed which required some some changes in 7dbf06c33508820486eed59aba78d852c6212864

pablomartin4btc reviewed (commented):
re-ACK 1ffea3fba3b22c40bbcc44cd7c3a637e0450315a

Since my last review:
- commit "test: rpcs disabled for descriptor wallets were removed" was moved to #28710
- addressed feedback from @rkrux

pablomartin4btc commented:
`-legacy` argument shouldn't be removed from `bitcoin-wallet`/ `wallettool.cpp`? (leaving `-descriptors` `true` by default):

```
/build/bin/bitcoin-wallet -legacy -wallet="pepe" create
Topping up keypool...
Wallet info
===========
Name: pepe
Format: sqlite
Descriptors: no
Encrypted: no
HD (hd seed available): yes
Keypool Size: 2000
Transactions: 0
Address Book: 0

```

achow101 committed:
wallet: Disallow legacy wallet creation from the wallet tool

achow101 commented:
> `-legacy` argument shouldn't be removed from `bitcoin-wallet`/ `wallettool.cpp`? (leaving `-descriptors` `true` by default):

Good catch! Added a commit to disable legacy wallet creation from the wallet tool.

laanwj reviewed (approved):
Re-ACK e39118ab0bf08fdaf68b16c86ad304b9133b43c7
git range-diff b8cefeb221490868d62b7a0695f8b5ea392d3654..545a36ae0aa5687366fc4fdd54058072f333e73f b2bb27f40c7ab7b51720e41d136dcc9077a525b3..e39118ab0bf08fdaf68b16c86ad304b9133b43c7 commit
- `test: rpcs disabled for descriptor wallets were removed` was dropped and moved to another PR
- `tests, gui: Use descriptors watchonly wallet for watchonly test` was added
- `test: Remove legacy wallet unit tests`: qt related changes were removed, replaced by the above commit
- `test: remove legacy wallet functional tests`: comment improvement
- `wallet: Disallow legacy wallet creation from the wallet tool` was added
- `wallet: Disallow loading legacy wallets`:  createwallet RPC error message improvement, remove redundant USE_BDB gated code
- `wallet: Disallow loading legacy wallets`: set format for new databases, remove redundant comment

Sjors commented on src/wallet/wallettool.cpp:
```diff
@@ -115,13 +115,25 @@ bool ExecuteWalletToolFunc(const ArgsManager& args, const std::string& command)
+            return false;
+        }
+        if (args.GetBoolArg("-legacy", true)) {
+            tfm::format(std::cerr, "The -legacy option must be set to \"false\"");
```
In ec5d2c7fca5e6c36aa59aeb7ce64b98739b6d094 "wallet: Disallow legacy wallet creation from the wallet tool": I'm getting a trailing `%` on macOS

davidgumberg commented on test/functional/wallet_backwards_compatibility.py:
```diff
@@ -325,7 +316,7 @@ def run_test(self):
             wallet.unloadwallet()
 
-            # Check that no automatic upgrade broke downgrading the wallet
+            # Check that no automatic upgrade broke the downgrading the wallet
```
Nit: Extra "the".

achow101 replied:
Fixed

Sjors commented:
re-ACK 17bb63f9f9b08e6af60c089234fe878657dbc88e

pablomartin4btc reviewed (commented):
re-ACK 17bb63f9f9b08e6af60c089234fe878657dbc88e

Since my last review:
- > Added a commit to disable legacy wallet creation from the wallet tool.
- Addresed @davidgumberg's feedback.

laanwj reviewed (approved):
re-ACK 17bb63f9f9b08e6af60c089234fe878657dbc88e
git range-diff b2bb27f40c7ab7b51720e41d136dcc9077a525b3..e39118ab0bf08fdaf68b16c86ad304b9133b43c7 9a4c92eb9ac29204df3d826f5ae526ab16b8ad65..17bb63f9f9b08e6af60c089234fe878657dbc88e
- `test: wallet_transactiontime_rescan importdescriptors always rescans`: commit removed (test change not necessary here)
- `test: Remove legacy wallet tests from wallet_backwards_compatibility.py`: comment change typo reverted
- `wallet: Disallow legacy wallet creation from the wallet tool`, `wallet: Disallow creating legacy wallets`: add newlines to `tfm::format`s 
  - (ideally these would both be in the first commit as that's where they're introduced, maybe if you need to re-touch anyway)

pablomartin4btc commented:
If a user tries to restore a legacy wallet (using RPC or QT) setting "load_on_startup" (can't be done on QT but it's being set in the wallet interface code), next time the node or QT starts it will be closed with the error _"... Failed to open database path ... The wallet appears to be a Legacy wallet, please use the wallet migration tool... "_. That case shouldn't be handled here? We shouldn't allow `load_on_startup` on legacy...

(Currently this situation is not happening as a side effect of [#31451](https://github.com/bitcoin/bitcoin/pull/31451#issuecomment-2828134729)).

achow101 commented:
> If a user tries to restore a legacy wallet (using RPC or QT) setting "load_on_startup" (can't be done on QT but it's being set in the wallet interface code), next time the node or QT starts it will be closed with the error _"... Failed to open database path ... The wallet appears to be a Legacy wallet, please use the wallet migration tool... "_. That case shouldn't be handled here? We shouldn't allow `load_on_startup` on legacy...
> 
> (Currently this situation is not happening as a side effect of [#31451](https://github.com/bitcoin/bitcoin/pull/31451#issuecomment-2828134729)).

I don't quite follow what you're saying. I tried to restore a legacy wallet with this PR and it doesn't do it, nor are there any issues with startup the next time.

Do you mean there is a possibility here if the interface is used incorrectly internally? In that case, I think it can be handled in a followup/if it becomes a reachable problem.

pablomartin4btc commented:
> I tried to restore a legacy wallet with this PR and it doesn't do it, nor are there any issues...

Ok, I thought the restore would allow to do it and then the load_on_startup would fail but the restore fail on the wallet file verification, so all good.

Tested `restorewallet` creating a legacy on `master`, backed it up there and trying to restore it on this PR's branch.
```
./build_31250/bin/bitcoin-cli -regtest -datadir=/tmp/btc-wallet restorewallet "restored_legacy_master" /tmp/btc-wallet/regtest/wallets/backup_legacy_master.dat true
error code: -18
error message:
Wallet file verification failed. Failed to open database path '/tmp/btc-wallet/regtest/wallets/restored_legacy_master'. The wallet appears to be a Legacy wallet, please use the wallet migration tool (migratewallet RPC).

```

pablomartin4btc commented:
> Do you mean there is a possibility here if the interface is used incorrectly internally?

I've tested it with this branch's QT and same failure on wallet file verifitcation while trying to restore a legacy backup using the GUI. All good.

fanquake commented:
Looks like `is_bdb_compiled` & `skip_if_no_bdb` should have been dropped.

pablomartin4btc commented:
> Looks like `is_bdb_compiled` & `skip_if_no_bdb` should have been dropped.

That's done in #28710 (commit: "_build, wallet, doc: Remove BDB_").

achow101 commented on test/functional/test_framework/test_node.py:
```diff
@@ -926,22 +925,16 @@ def send_cli(self, clicommand=None, *args, **kwargs):
-
     def importprivkey(self, privkey, label=None, rescan=None):
         wallet_info = self.getwalletinfo()
         if 'descriptors' not in wallet_info or ('descriptors' in wallet_info and not wallet_info['descriptors']):
```
The `importprivkey` overload was kept because several tests still use it.

achow101 commented on test/functional/rpc_createmultisig.py:
```diff
@@ -25,14 +25,10 @@
         self.setup_clean_chain = True
         self.num_nodes = 3
         self.supports_cli = False
-        self.enable_wallet_if_possible()
```
Added a commit to #28710 to remove

achow101 commented on test/functional/rpc_createmultisig.py:
```diff
@@ -42,29 +38,20 @@ def create_keys(self, num_keys):
-                self.do_multisig(keys, sigs, output_type, wallet_multi)
+                self.do_multisig(keys, sigs, output_type)
 
-        self.test_multisig_script_limit(wallet_multi)
```
Added a commit to #28710 to restore

achow101 commented on test/functional/test_framework/test_node.py:
```diff
@@ -926,22 +925,16 @@ def send_cli(self, clicommand=None, *args, **kwargs):
 
 class RPCOverloadWrapper():
-    def __init__(self, rpc, cli=False, descriptors=False):
+    def __init__(self, rpc, cli=False):
```
I'll let the next person to touch it to change this.

achow101 commented on test/functional/test_runner.py:
```diff
@@ -151,71 +138,54 @@
     'wallet_txn_doublespend.py --mineblock',
     'tool_bitcoin_chainstate.py',
-    'tool_wallet.py --legacy-wallet',
-    'tool_wallet.py --legacy-wallet --bdbro',
-    'tool_wallet.py --legacy-wallet --bdbro --swap-bdb-endian',
```
They are removed in #28710

achow101 commented on test/functional/wallet_createwallet.py:
```diff
@@ -156,12 +143,12 @@ def run_test(self):
         resp = self.nodes[0].createwallet(wallet_name='w7', disable_private_keys=False, blank=False, passphrase='')
-        assert_equal(resp["warnings"], [EMPTY_PASSPHRASE_MSG] if self.options.descriptors else [EMPTY_PASSPHRASE_MSG, LEGACY_WALLET_MSG])
+        assert_equal(resp["warnings"], [EMPTY_PASSPHRASE_MSG])
 
```
Added a commit to #28710 to remove

achow101 commented on test/functional/feature_bip68_sequence.py:
```diff
@@ -58,6 +55,7 @@ def set_test_params(self):
                 '-testactivationheight=csv@432',
             ],
         ]
+        self.uses_wallet = None
```
added a commit to #28710 to remove

amtriorix commented:
I do disagree with this merge
Are you crazy or what.  Backwards compatibility is needed for wallet.dat

davidgumberg commented:
> Backwards compatibility is needed for wallet.dat

Bitcoin Core will still be able to migrate wallets that are in the legacy format to the new format. See the `migratewallet` rpc command: https://bitcoincore.org/en/doc/29.0.0/rpc/wallet/migratewallet/ and the legacy wallet deprecation tracking issue (https://github.com/bitcoin/bitcoin/issues/20160) for more context.
